new Text analyzers page + Lang analyzer section (#1376)
opensearch-project/documentation-website: _opensearch/query-dsl/text-analyzers.md

---
layout: default
title: Text analyzers
parent: Query DSL
nav_order: 41
---

# Text analyzers

OpenSearch applies text analysis during indexing or searching for `text` fields. There is a standard  analyzer that OpenSearch uses by default for text analysis. To optimize unstructured text for search, you can convert it into structured text with our text analyzers.

## Text analyzers

OpenSearch provides several text analyzers to convert your structured text into the format that works best for your searches.

OpenSearch supports the following text analyzers:

> 1. **Standard analyzer** – parses all strings into terms at word boundaries per the Unicode text segmentation algorithm. It removes most, but not all punctuation. It converts all characters to lowercase. You can remove stop words if you turn on that option, but it does not remove stop words by default.

Reviewer: ```suggestion
1. **Standard analyzer** – parses strings into terms at word boundaries per the Unicode text segmentation algorithm. It removes most, but not all punctuation. It converts strings to lowercase. You can remove stop words if you turn on that option, but it does not remove stop words by default.
```

> 1. **Standard analyzer** – parses strings into terms at word boundaries per the Unicode text segmentation algorithm. It removes most, but not all punctuation. It converts strings to lowercase. You can remove stop words if you turn on that option, but it does not remove stop words by default.

Reviewer 2: All list items: Capitalize the first word following the en dash. Insert a comma after "all".

> 1. **Simple analyzer** – converts all characters to lowercase, and removes non-letter characters when it splits a string into tokens on any non-letter character.

Reviewer: ```suggestion
1. **Simple analyzer** – converts strings to lowercase, and removes non-letter characters when it splits a string into tokens on any non-letter character.
```

> 1. **Simple analyzer** – converts strings to lowercase, and removes non-letter characters when it splits a string into tokens on any non-letter character.

Reviewer 2: Remove the comma.

> 1. **Whitespace analyzer** – parses a string into terms between each whitespace.

Reviewer: ```suggestion
1. **Whitespace analyzer** – parses strings into terms at whitespace characters.
```

> 1. **Stop analyzer** – Converts a string to lowercase and removes non-letter characters by splitting the string into tokens at each non-letter character. It also removes stop words (e.g. "but," or "this") from the query string.

Reviewer: ```suggestion
1. **Stop analyzer** – Converts strings to lowercase and removes non-letter characters by splitting strings into tokens at each non-letter character. It also removes stop words (e.g. "but," or "this") from strings.
```

> 1. **Stop analyzer** – Converts strings to lowercase and removes non-letter characters by splitting strings into tokens at each non-letter character. It also removes stop words (e.g. "but," or "this") from strings.

Reviewer 2: Insert a comma after "e.g.". Remove the comma after "but".

1. **Keyword analyzer** – receives a string as input and outputs the entire string as one term.

> 1. **Pattern analyzer** – splits a string into terms using a regular expression and supports converting characters to lowercase. It also supports removing stop words.

Reviewer: ```suggestion
1. **Pattern analyzer** – splits strings into terms using regular expressions and supports converting strings to lowercase. It also supports removing stop words.
```

1. **Language analyzer** – provides analyzers specific to multiple languages.
1. **Fingerprint analyzer** – creates a fingerprint to use as a duplicate detector.

> The full specialized text analyzers reference is in-progress and will be published soon.

Reviewer 2: Remove the hyphen from "in-progress".

{: .note }

## How to use text analyzers

> If you want to use a text analyzer, specify the name of the analyzer for the `analyzer` field: standard, simple, whitespace, stop, keyword, pattern, fingerprint, and language.

Reviewer 2: It looks like "and" should be "or".

> Each analyzer consists of one tokenizer and zero or more token filters. Different analyzers have different character filters, tokenizers and token filters. To pre-process the string before the tokenizer is applied, you can use one or more character filters.

Reviewer 2: Insert the serial comma after "tokenizers".

---
layout: default
title: Text analyzers
parent: Query DSL
nav_order: 41
---

> # Text analyzers

Reviewer: Do we just want to call this section "Analyzers" and include a sentence that mentions that analysis is only used on `text` fields?

OpenSearch applies text analysis during indexing or searching for `text` fields. There is a standard  analyzer that OpenSearch uses by default for text analysis. To optimize unstructured text for search, you can convert it into structured text with our specialized text analyzers.

> ## Specialized analyzers

Reviewer: Add a newline before this line

Reviewer: I'm not sure the modifier "specialized" is needed here.

Author: We are differentiating ourselves from Elastic's "built-in" analyzers. We also have these, they are called "specialized" analyzers.

OpenSearch provides specialized analyzers to convert your structured text into the format that works best for your searches.

OpenSearch supports the following specialized text analyzers:

> 1. **Standard analyzer** – parses text strings into terms at word boundaries per the Unicode text segmentation algorithm. It also removes stop words and removes punctuation and lowercase terms.

Reviewer: Standard analyzer does not remove stop words by default. It **can** remove stop words if you turn on an option on the analyzer. It removes most but not all punctuation. It also converts all characters to lower case.

Reviewer: I think you meant "removes punctuation and lowercase**s** terms"

Author: thanks for clarifying info

> 1. **Simple analyzer** – converts all characters to lower case and divides groups of non-letter characters into terms.

Reviewer: "divides groups of non-letter characters into terms" - this is not entirely true. It splits a string into tokens on any non-letter character, removing the non-letter characters.

> 1. **Whitespace analyzer** – parses characters into terms between whitespaces.

Reviewer: Maybe "parses a string into terms on whitespace characters"?

> 1. **Stop analyzer** – Removes stop words (e.g. "but," or "this") from the query string.

Author: ok thanks

> 1. **Keyword analyzer** – receives text and outputs only the text specified as a single keyword term.

Reviewer: returns the whole input string as one term

> 1. **Pattern analyzer** – splits text into terms using a regular expression and supports lower case characters and stop words.

Reviewer: ... supports lowercasing the string and removing stop words

Author: we can't say "lowercasing" that's not a verb.

> 1. **Language analyzer** – provides multiple languages to specify as analyzer values.

Reviewer: provides analyzers specific to multiple languages

1. Fingerprint analyzer – creates a fingerprint to use as a duplicate detector.

> 

Reviewer: Do we also want to mention here that you can create a custom analyzer?

Reviewer: So, the above are built-in, but if you want to create your own, you can.

Author: Sounds good. Please provide that information about how to create a custom analyzer. We can iterate and add that to the next PR.

The full specialized text analyzers reference is in-progress and will be published soon.
{: .note }

## How to use specialized analyzers

If you want to use a specialized text analyzer, you specify the name of the analyzer for the `analyzer` field.

> You can use the following options with the `analyzer` field: standard, simple, whitespace, stop, keyword, pattern, fingerprint, and language. Different analyzers have different character filters, tokenizers, and token filters.

Reviewer: Is `with` the right preposition here? We want to convey that the `analyzer` field takes these options, and also `custom` or `default`, which we should describe here.
Second sentence: I think we have to be more specific here: an analyzer consists of one tokenizer and zero or more token filters. To preprocess the string before the tokenizer is applied, you can use one or more character filters.

Author: I simplified it with exact "specify the name of the analyzer for the analyzer field" and gave an example. 
Thanks for tokenizer description, I've added your description.
